src/app/clusters/mode-base-server/mode-base-server.h

```diff
@@ -0,0 +1,198 @@
+     */
+    bool HasFeature(Feature feature) const;
+
+    /**
+     * This function returns true if the mode value given matches one of the supported modes, otherwise it returns false.
+     * @param mode
+     */
+    bool IsSupportedMode(uint8_t mode);
+
+private:
+    EndpointId mEndpointId{};
+    ClusterId mClusterId{};
+
+    // Attribute data store
+    uint8_t mCurrentMode;
+    DataModel::Nullable<uint8_t> mStartUpMode;
+    DataModel::Nullable<uint8_t> mOnMode;
+
+    uint32_t mFeature;
+
+    /**
+     * This checks to see if this clusters instance is a valid ModeBase aliased cluster based on the AliasedClusters list.
+     * @return true if the clusterId of this instance is a valid ModeBase cluster.
+     */
+    bool isAliasCluster() const;
+
+    /**
+     * Internal change-to-mode command handler function.
+     */
+    void handleChangeToMode(HandlerContext & ctx, const Commands::ChangeToMode::DecodableType & req);
+
+public:
+    /**
+     * Creates a mode base cluster instance. The Init() function needs to be called for this instance to be registered and
+     * called by the interaction model at the appropriate times.
+     * @param aEndpointId The endpoint on which this cluster exists. This must match the zap configuration.
+     * @param aClusterId The ID of the ModeBase aliased cluster to be instantiated.
+     * @param aFeature The bitmask value that identifies which features are supported by this instance.
+     */
+    Instance(EndpointId aEndpointId, ClusterId aClusterId, uint32_t aFeature) :
+        CommandHandlerInterface(Optional<EndpointId>(aEndpointId), aClusterId),
+        AttributeAccessInterface(Optional<EndpointId>(aEndpointId), aClusterId)
+    {
+        mEndpointId = aEndpointId;
+        mClusterId  = aClusterId;
+        mFeature    = aFeature;
+    }
+
+    ~Instance() override { ModeBaseAliasesInstanceMap.erase(mClusterId); }
+
+    template <typename RequestT, typename FuncT>
+    void HandleCommand(HandlerContext & handlerContext, FuncT func);
+
+    // The following functions should be overridden by the SDK user to implement the business logic of their application.
+
+    /**
+     * This init function will be called during the ModeBase server initialization after the Instance information has been
+     * validated and the Instance has been registered. This can be used to initialise app logic.
+     */
+    virtual CHIP_ERROR AppInit() = 0;
+
+    /**
+     * Returns the number of modes managed by this instance.
+     */
+    virtual uint8_t NumberOfModes();
+
+    /**
+     * Get the mode label of the Nth mode in the list of modes.
+     * @param modeIndex The index of the mode to be returned. It is assumed that modes are indexable from 0 and with no gaps.
+     * @param label a reference to the MutableCharSpan that is to contain the mode label. Use CopyCharSpanToMutableCharSpan
+     * to copy into the MutableCharSpan.
+     * @return Returns a CHIP_NO_ERROR if there was no error.
```

Comment:
What errors can arise?

Reply:
The errors here depend on the implementation by the SDK consumer. Typical errors are `CHIP_ERROR_NOT_FOUND` if the mode does not exist and `CHIP_ERROR_INVALID_ARGUMENT` if the input variable to be mutated is incorrect. Do you think we should document something like this here?

Reply:
What should be documented here are any special error values the cluster implementation depends on, if any.

Reply:
Following Boris' suggestion, I have change the API for these functions and documented the expected errors as
```
 * @return Returns a CHIP_NO_ERROR if there was no error and the label was returned successfully.
 * CHIP_ERROR_PROVIDER_LIST_EXHAUSTED if the modeIndex in beyond the list of available labels.
```